util/flask_util.py

```diff
@@ -16,4 +22,125 @@ def problem(type, status, title, detail=None, instance=None, headers={}):
     """Create a Response that includes a Problem Detail Document."""
     status, headers, data = problem_raw(
         type, status, title, detail, instance, headers)
-    return Response(data, status, headers)
+    return FlaskResponse(data, status, headers)
+
+
+class Response(FlaskResponse):
+    """A Flask Response object with some conveniences added.
+
+    The conveniences:
+
+       * It's easy to calculate header values such as Cache-Control.
+       * A response can be easily converted into a string for use in
+         tests.
+    """
+
+    def __init__(self, response=None, status=None, headers=None, mimetype=None,
+                 content_type=None, direct_passthrough=False, max_age=0,
+                 private=None):
+        """Constructor.
+
+        All parameters are the same as for the Flask/Werkzeug Response class,
+        with these additions:
+
+        :param max_age: The number of seconds for which clients should
+            cache this response. Used to set a value for the
+            Cache-Control header.
+        :param private: If this is True, then the response contains
+            information from an authenticated client and should not be stored
+            in intermediate caches.
+        """
+        max_age = max_age or 0
+        try:
+            max_age = int(max_age)
+        except ValueError:
+            max_age = 0
+        self.max_age = max_age
+        if private is None:
+            if self.max_age == 0:
+                # The most common reason for max_age to be set to 0 is that a resource
+                # is _also_ private.
+                private = True
+            else:
+                private = False
+        self.private = private
+
+        body = response
+        if isinstance(body, etree._Element):
+            body = etree.tostring(body)
+        elif not isinstance(body, (bytes, unicode)):
+            body = unicode(body)
+
+        super(Response, self).__init__(
+            response=body,
+            status=status,
+            headers=self._headers(headers or {}),
+            mimetype=mimetype,
+            content_type=content_type,
+            direct_passthrough=direct_passthrough
+        )
+
+    def __unicode__(self):
+        """This object can be treated as a string, e.g. in tests.
+
+        :return: The entity-body portion of the response.
+        """
+        return self.data
+
+    def _headers(self, headers={}):
+        """Build an appropriate set of HTTP response headers."""
+        # Don't modify the underlying dictionary; it came from somewhere else.
+        headers = dict(headers)
+
+        # Set Cache-Control based on max-age.
+        if self.private:
+            private = "private"
+        else:
+            private = "public"
+        if self.max_age and isinstance(self.max_age, int):
+            # A CDN should hold on to the cached representation only half
```

Comment:
Why is this the case?

Reply:
CDN cache times and client cache times are additive. Let's say the CDN cache time is 30 minutes and the client cache time is 30 minutes. When the client requests a document, it might get a document that's 29 minutes out of date. Then it holds on to the document for 30 minutes. At the end of this process, the client is using a document that's nearly an hour out of date, even though the cache time is only 30 minutes.

But the whole point of the CDN cache is to stop client traffic from hitting the origin server. So this problem is unavoidable if we want the CDN to be useful. It's a tradeoff we have to figure out how to make. Setting the CDN cache time to 1/2 of the client cache time is a rule of thumb that helps mitigate the problem while keeping the CDN useful. There's nothing special about the 1/2 rule, but I don't have any ideas for coming up with a better coefficient.